perma_web/fabfile/dev.py

```diff
@@ -573,3 +573,113 @@ def check_s3_hashes():
+    if hasattr(default_storage, 'secondary_storage'):
+        storages.update({'secondary': {'storage': default_storage.secondary_storage, 'lookup': {}}})
+
+    # only use cache files when all are present: link cache, and one for each storage
+    link_cache = '/tmp/perma_link_cache.txt'
+    caches = [link_cache]
+    for key in storages:
+        caches.append('/tmp/perma_storage_cache_{0}.txt'.format(key))
+
+    if not all(map(os.path.exists, caches)):
+        print("Building link cache ...")
+        with open(link_cache, 'w') as tmp_file:
+            capture_filter = (Q(role="primary") & Q(status="success")) | (Q(role="screenshot") & Q(status="success"))
+            for link in tqdm(Link.objects.filter(captures__in=Capture.objects.filter(capture_filter)).distinct()):
+                tmp_file.write("{0}\n".format(link.warc_storage_file()))
+                # this produces strings like u'warcs/0G/GO/XR/XG/0-GGOX-RXGQ.warc.gz'; make the storage paths match
+                # by chopping off the prefix, whether storage.location, ._root_path, or .base_location
+        print("Building storage cache{0} ...".format("s" if len(storages) > 1 else ""))
+        for key in storages:
+            storage = storages[key]['storage']
+            with open('/tmp/perma_storage_cache_{0}.txt'.format(key), 'w') as tmp_file:
+                if hasattr(storage, 'bucket'):
+                    # S3
+                    for f in tqdm(storage.bucket.list('generated/warcs/')):
+                        # etag is a string like u'"3ea8c903d9991d466ec437d1789379a6"', so we need to
+                        # knock off the extra quotation marks
+                        # note that etags are not always md5sums, but should be in these cases; we can rewrite
+                        # or replace md5hash if necessary
+                        path = f.key[(len(storage.location)):]
+                        hash = f.etag[1:-1]
+                        tmp_file.write("{0}\t{1}\n".format(path, hash))
+                        storages[key]['lookup'][path] = hash
+                else:
+                    if hasattr(storage, '_root_path'):
+                        # SFTP
+                        base = storage._root_path
+                    else:
+                        # local file storage -- are there other possibilities to consider?
+                        base = storage.base_location
+                    for f in tqdm(storage.walk(os.path.join(base, 'warcs'))):
+                        # os.walk: "For each directory in the tree rooted at directory top (including top itself),
+                        # it yields a 3-tuple (dirpath, dirnames, filenames)" -- there should be exactly one WARC
+                        # per directory, so:
+                        if len(f[2]) == 1:
+                            full_path = os.path.join(f[0], f[2][0])
+                            path = full_path[len(base):]
```

Comment:
Same here: a duplicative comment might not be a bad thing here, just to make it clearer what's happening where. Or not.